docs(permissions): add guide for managing and restricting control plane permissions (#2305)
kumahq/kuma-website: app/_src/guides/restrict-permissions-to-selected-namespaces-on-kubernetes.md

---

> title: Restrict permissions to selected namespaces on Kubernetes

Reviewer: This isn't the guide we discussed no?

I thought we said:

1. install with restricted to just 1 ns
2. install the demo there
3. show it works
4. install the demo in another namespace
5. show it doesn't work
6. reinstall with the 2nd namespace added
7. restart the workload in the 2nd namespace
8. show it works

This ^ is most likely what users will actually do.

Author: done

{% tabs codeblock %}
{% tab kumactl %}
```bash
kumactl install control-plane | kubectl apply -f -
```
{:.no-line-numbers}
{% endtab %}
{% tab Helm %}
```bash
# Before installing {{ Kuma }} with Helm, configure your local Helm repository:
# {{ site.links.web }}{{ docs }}/production/cp-deployment/kubernetes/#helm
helm install \
  --create-namespace \
  -n {{ site.mesh_namespace }} \
  {{ site.mesh_helm_install_name }} {{ site.mesh_helm_repo }}
```
{:.no-line-numbers}
{% endtab %}
{% endtabs %}

## Step 2: Deploy demo applications

Next, deploy demo applications in two separate namespaces, `kuma-demo` and `kuma-demo-other`:

```bash
kumactl install demo --without-gateway -n kuma-demo | kubectl apply -f -
kumactl install demo --without-gateway -n kuma-demo-other | kubectl apply -f -
```
{:.no-line-numbers}

## Step 3: Verify setup in the GUI

Port-forward the control plane GUI by running:

```bash
kubectl port-forward svc/{{ site.mesh_cp_name }} -n {{ site.mesh_namespace }} 5681:5681
```
{:.no-line-numbers}

Then, open your browser at [http://localhost:5681/gui](http://localhost:5681/gui):

* Click **Meshes** on the sidebar.
* Select the **default** mesh.

> * Go to the **Services** tab.

Author: addressed

---
title: Restrict permissions to selected namespaces on Kubernetes
content_type: tutorial
---

{% capture docs %}/docs/{{ page.release }}{% endcapture %}
{% assign Kuma = site.mesh_product_name %}
{% assign kuma = site.mesh_install_archive_name | default: "kuma" %}
{% assign kuma-control-plane-workloads = kuma | append: "-control-plane-workloads" %}

By default, {{ Kuma }} deployed on Kubernetes has permissions to observe and react to events from resources across the entire cluster. While this behavior simplifies initial setup and testing, it might be too permissive for production environments. Limiting {{ Kuma }}'s access to only necessary namespaces helps enhance security and prevents potential impact on unrelated applications.

> Starting from version 2.11, you can restrict the permissions granted to the {{ Kuma }} control plane. This guide explains how to limit {{ Kuma }} to specific namespaces, giving you greater control over security and resource management.

Reviewer 2: ```suggestion
This guide explains how to limit {{ Kuma }} to specific namespaces, giving you greater control over security and resource management.
```

dosc should be current state of project, we should not say when we've added something. This doc won't be visible in older versions and in newer it does not matter

Author: addressed

> ## Prerequisites

Reviewer 2: we could have same preerequisites as in quickstart: https://kuma.io/docs/2.10.x/quickstart/kubernetes-demo/#prerequisites

Author: addressed

Before you start, ensure you have:

> * A clean, running, and accessible Kubernetes cluster

Reviewer 2: can we just add step with:
```
minikube start -p mesh-zone
```

as we do in quickstart?

Author: addressed

* The `kubectl` command-line tool installed and configured

## Restrict {{ Kuma }} to selected namespaces

This section shows you how to restrict {{ Kuma }}'s permissions to selected namespaces, verify that restrictions are correctly applied, and update permissions if needed.

### Step 1: Create and label the initial namespace

```bash
kubectl create namespace first-namespace
kubectl label namespace first-namespace kuma.io/sidecar-injection=enabled
```

### Step 2: Install {{ Kuma }} restricted to the first namespace

{% cpinstall namespaceAllowList %}
namespaceAllowList={first-namespace}
{% endcpinstall %}

### Step 3: Deploy a test workload in the allowed namespace

```bash
kubectl run nginx --image=nginx --port=80 --namespace first-namespace
```

### Step 4: Verify the first namespace is working

Check that the control plane is managing the workload in `first-namespace`:

```bash

> kubectl get dataplanes --namespace first-namespace

Author: addressed

kubectl get pods --namespace first-namespace
kubectl get rolebindings --namespace first-namespace
```

You should see:

* A `Dataplane` resource listed
* The pod showing two containers (`2/2`), one for `nginx` and one for the sidecar

> * A `RoleBinding` named `{{ kuma-control-plane-workloads }}` that grants the control plane elevated access to manage resources within the namespace

Reviewer: Technically because the pod has a defined name now you could just show the output

Author: addressed

### Step 5: Create and label a second namespace

```bash
kubectl create namespace second-namespace
kubectl label namespace second-namespace kuma.io/sidecar-injection=enabled
```

### Step 6: Deploy the same workload in the second namespace

```bash
kubectl run nginx --image=nginx --port=80 --namespace second-namespace
```

### Step 7: Verify the second namespace is not working

```bash
kubectl get dataplanes --namespace second-namespace

> kubectl get pods --namespace second-namespace

Author: addressed

```

No `Dataplanes` should be present, and pods should show only one container (`1/1`).

### Step 8: Update {{ Kuma }} to include the second namespace

{% cpinstall namespaceAllowListMore %}
namespaceAllowList={first-namespace,second-namespace}
{% endcpinstall %}

### Step 9: Restart workloads in the second namespace

Since the workload was created as a single pod (not part of a `Deployment`), delete the existing pod and run it again:

```bash
kubectl delete pod --namespace second-namespace --all
kubectl run nginx --image=nginx --port=80 --namespace second-namespace
```

### Step 10: Verify the second namespace is now working

```bash

> kubectl get dataplanes --namespace second-namespace

Author: addressed

kubectl get pods --namespace second-namespace
kubectl get rolebindings --namespace second-namespace
```

Just like with the first namespace, you should see:

* A `Dataplane` resource
* The pod showing two containers (`2/2`)
* A `RoleBinding` named `{{ kuma-control-plane-workloads }}`

> ## Cleanup

Author: addressed

---
title: Restrict permissions to selected namespaces on Kubernetes
content_type: tutorial
---

{% capture docs %}/docs/{{ page.release }}{% endcapture %}
{% assign Kuma = site.mesh_product_name %}
{% assign kuma = site.mesh_install_archive_name | default: "kuma" %}

> {% assign kuma-demo = kuma | append: "-demo" %}

Reviewer: In KM kuma-demo remains kuma-demo no?

Author: it's irrelevant anymore after the latest changes

{% assign kuma-another-demo = kuma | append: "-another-demo" %}

> By default, {{ Kuma }} deployed on Kubernetes has permission to observe and react to events from resources across the entire cluster. While this behavior simplifies initial setup and testing, it might be too permissive for production environments. Limiting {{ Kuma }}'s access to only necessary namespaces helps enhance security and prevents potential impact on unrelated applications.

Reviewer: ```suggestion
By default, {{ Kuma }} deployed on Kubernetes has permissions to observe and react to events from resources across the entire cluster. While this behavior simplifies initial setup and testing, it might be too permissive for production environments. Limiting {{ Kuma }}'s access to only necessary namespaces helps enhance security and prevents potential impact on unrelated applications.
```

Author: addressed

Starting from version 2.11, you can restrict the permissions granted to the {{ Kuma }} control plane. This guide explains how to limit {{ Kuma }} to specific namespaces, giving you greater control over security and resource management.

## Prerequisites

Before you start, ensure you have:

* A clean, running, and accessible Kubernetes cluster
* The `kubectl` command-line tool installed and configured
* The `kumactl` command-line tool installed

> 

Reviewer: Usually when you get into list items that go higher than 5 it indicates you are going to loose the reader.

Here I suggest a way to reorganise this in subsections where each section is a logical end.
What you get from these sections is that they have each a few items which makes it easier to track and makes the tuto look less like an insurance contract.

Author: Do you mean amount of steps in this case?

Reviewer: Yes create sub titles with at most 5 items in each.

> ## Restrict {{ Kuma }} to selected namespaces

Reviewer: ```suggestion
## Install {{ Kuma }} a restrict it to `kuma-demo` namespace
```

Author: addressed

Follow these steps:

{% capture namespaceAllowList %}
{% cpinstall namespaceAllowList %}
namespaceAllowList={% raw %}{{% endraw %}{{ kuma-demo }}{% raw %}}{% endraw %}
{% endcpinstall %}
{% endcapture %}

{% capture namespaceAllowListMore %}
{% cpinstall namespaceAllowListMore %}
namespaceAllowList={% raw %}{{% endraw %}{{ kuma-demo }},{{ kuma-another-demo }}{% raw %}}{% endraw %}
{% endcpinstall %}
{% endcapture %}

{% capture cleanup %}
{% tabs codeblock %}
{% tab kumactl %}
```bash
kumactl install control-plane \
  --set "{{ site.set_flag_values_prefix }}namespaceAllowList={% raw %}{{% endraw %}{{ kuma-demo }},{{ kuma-another-demo }}{% raw %}}{% endraw %}" \
  | kubectl delete -f -
```
{% endtab %}
{% tab Helm %}
```bash
helm uninstall {{ kuma }} -n {{ kuma }}-system
```
{% endtab %}
{% endtabs %}
{% endcapture %}

1. **Install {{ Kuma }} restricted initially to just the `{{ kuma-demo }}` namespace**

   {{ namespaceAllowList | indent }}

2. **Deploy a demo application into the allowed namespace**

   ```bash
   kumactl install demo -n {{ kuma-demo }} | kubectl apply -f -

>    ```

Reviewer: This also avoids any requirement of kumactl which is a GREAT thing :)

>    ```

Author: addressed

3. **Verify the demo is working correctly in the allowed namespace**

   You should see `Dataplanes` created and pods with sidecar injection:

   ```bash
   kubectl get dataplanes -n {{ kuma-demo }}
   kubectl get pods -n {{ kuma-demo }}
   ```

   Pods should show two containers each (`2/2`).

> 4. **Deploy another demo application in a different namespace (`{{ kuma-another-demo }}`), which isn't yet allowed**

Reviewer: ```suggestion
## Ensure that dataplanes are not created when deploying in a new namespace
```

   ```bash

>    kumactl install demo -n {{ kuma-another-demo }} | kubectl apply -f -

Author: addressed

   ```

   Confirm sidecar injection is enabled:

   ```bash
   kubectl get namespace {{ kuma-another-demo }} -o yaml
   ```

>    Returned namespace should have an annotation:

Reviewer: ```suggestion
   Returned namespace should have a label:
```

Author: addressed

   ```yaml
   kuma.io/sidecar-injection: enabled
   ```

5. **Verify that the second namespace is not yet working**

   `Dataplanes` won't appear, pods will have only one container, and no `RoleBindings` will exist.

   ```bash
   kubectl get dataplanes -n {{ kuma-another-demo }}
   kubectl get pods -n {{ kuma-another-demo }}
   kubectl get rolebindings -n {{ kuma-another-demo }}
   ```

> 6. **Update the {{ Kuma }} installation to include the second namespace**

Reviewer: ```suggestion
## Update the {{ Kuma }} installation to include a second `kuma-another-demo`
```

Author: addressed

   {{ namespaceAllowListMore | indent }}

7. **Restart workloads in the second namespace to apply the changes**

   ```bash
   kubectl rollout restart deployment -n {{ kuma-another-demo }}
   ```

8. **Confirm the second namespace now works as expected**

   `Dataplanes` should appear, pods should have two containers, and `RoleBindings` should exist:

   ```bash
   kubectl get dataplanes -n {{ kuma-another-demo }}
   kubectl get pods -n {{ kuma-another-demo }}
   kubectl get rolebindings -n {{ kuma-another-demo }}
   ```

## Cleanup

When you're finished, clean up your environment with these steps:

1. **Remove demo applications and their namespaces**

   ```bash
   kumactl install demo -n {{ kuma-demo }} | kubectl delete -f -

>    kumactl install demo -n {{ kuma-another-demo }} | kubectl delete -f -

Author: addressed

This guide explains how to limit {{ Kuma }} to specific namespaces, giving you greater control over security and resource management.

## Prerequisites

Before you begin, make sure you have the following tools installed:

* [Helm](https://helm.sh/) – used to install and manage Kubernetes applications
* [minikube](https://minikube.sigs.k8s.io/docs/) – used to run a local Kubernetes cluster for testing

### Start a Kubernetes cluster

Start a local Kubernetes cluster using minikube. The `-p` flag creates a new profile named `mesh-zone`:

```bash
minikube start -p mesh-zone
```

{% tip %}
If you already have a running Kubernetes cluster, either locally or in the cloud (for example, EKS, GKE, or AKS), you can skip this step.
{% endtip %}

## Restrict {{ Kuma }} to selected namespaces

This section walks through configuring {{ Kuma }} to limit its access to specific namespaces. You’ll deploy test workloads, verify control plane behavior, and then expand access to additional namespaces.

> ### First namespace

Reviewer: ```suggestion
### Install Kuma which manages a single namespace
```

Author: addressed

```bash
kubectl get dataplanes --namespace first-namespace
```

Expected:

```
NAME    KUMA.IO/SERVICE             KUMA.IO/SERVICE
nginx   nginx_first-namespace_svc
```
{:.no-line-numbers}

Then check that the pod has the sidecar injected:

```bash
kubectl get pods --namespace first-namespace
```

Expected:

```
NAME    READY   STATUS    RESTARTS   AGE
nginx   2/2     Running   0          2m5s
```
{:.no-line-numbers}

Then verify the required RoleBinding:

```bash
kubectl get rolebindings --namespace first-namespace
```

Expected:

{% capture rbacHeader -%}
{%- if page.edition and page.edition != "kuma" %}
NAME                                ROLE                                            AGE
{%- else %}
NAME                           ROLE                                       AGE
{%- endif %}
{% endcapture %}

```
{{- rbacHeader -}}
{{ kuma-control-plane-workloads }}   ClusterRole/{{ kuma-control-plane-workloads }}   3m46s
```
{:.no-line-numbers}

This confirms that:

* A `Dataplane` was created
* The pod includes the `kuma-sidecar`
* A `RoleBinding` named `{{ kuma-control-plane-workloads }}` grants elevated access to the control plane

> ### Second namespace

Reviewer: ```suggestion
### Create a second namespace and check Kuma doesn't run in it
```

Author: addressed

```

#### Deploy the same workload in the second namespace

```bash
kubectl run nginx --image=nginx --port=80 --namespace second-namespace
```

#### Verify the second namespace is *not* working

Check that the control plane is **not** managing resources in `second-namespace`.

Run the following commands:

```bash
kubectl get dataplanes --namespace second-namespace
```

Expected output:

```
No resources found in second-namespace namespace.
```
{:.no-line-numbers}

This means no `Dataplane` was created.

```bash
kubectl get pods --namespace second-namespace
```

Expected output:

```
NAME    READY   STATUS    RESTARTS   AGE
nginx   1/1     Running   0          42s
```
{:.no-line-numbers}

This indicates the pod is running **without** the `kuma-sidecar`.

```bash
kubectl get rolebindings --namespace second-namespace
```

Expected output:

```
No resources found in second-namespace namespace.
```
{:.no-line-numbers}

This confirms that:

* The control plane does not have permission to manage this namespace
* The pod was started without sidecar injection
* No `RoleBinding` was created to grant control plane access

> #### Update {{ Kuma }} to include the second namespace

Reviewer: ```suggestion
### Update {{ Kuma }} to also manage the second namespace
```

Note the higher title

Author: addressed

> 

Reviewer: ```suggestion
#### Update Kuma to include `second-namespace`
```

Author: addressed
